src/cosim/utility/thread_pool.hpp

```diff
@@ -0,0 +1,106 @@
+/**
+*  \file
+*  Slave interface.
+*
+*  \copyright
+*      This Source Code Form is subject to the terms of the Mozilla Public
+*      License, v. 2.0. If a copy of the MPL was not distributed with this
+*      file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+#ifndef COSIM_UTILITY_THREAD_POOL_HPP
+#define COSIM_UTILITY_THREAD_POOL_HPP
+
+#include <atomic>
+#include <condition_variable>
+#include <functional>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace cosim
+{
+namespace utility
+{
+
+class thread_pool
+{
+private:
+    std::atomic_bool done_;
+    std::queue<std::function<void()>> work_queue_;
+    std::vector<std::thread> threads_;
+    std::mutex m_;
+    std::condition_variable cv_;
+
+    void worker_thread()
+    {
+        while (!done_) {
+
+            std::unique_lock<std::mutex> lck(m_);
+            if (!work_queue_.empty()) {
+
+                auto task = work_queue_.front();
+                work_queue_.pop();
+                task();
+
+                lck.unlock();
+                cv_.notify_one();
```

Comment:
No?

Reply:
The idea being to wake up one of the worker threads immediately after adding to the work queue, and in the same function, rather than waiting for one of the running threads to process the queue. 

This strikes me as a cleaner implementation that also removes the need to loop all the worker threads constantly. They can simply block until something calls `submit()`, then be notified from here.